src/main/java/com/amazon/ion/impl/bin/SymbolInliningStrategy.kt

```diff
@@ -0,0 +1,52 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.impl.bin
+
+/**
+ * A strategy to determine whether a SymbolToken with known text should be encoded by Symbol ID (SID) or as inline text.
+ * Symbols with unknown text are always be written as a SID because the text is unknown.
+ *
+ * Some possible implementation ideas:
+ *
+ *  - A simple implementation could elect to inline symbols that are less than `N` characters long.
+ *  - A domain-specific implementation could choose to inline symbols with specific prefixes. E.g. annotations starting
+ *    with `org.example` always get written inline.
+ *  - A stateful implementation could keep track of how often a symbol is used, and elect to write the symbol inline
+ *    until it has been used at least `N` times.
+ *  - A streaming-oriented implementation could keep track of the symbols that are used, and inline any symbols not
+ *    already in the symbol table. Once a top-level value is complete, some other component could inspect the list of
+ *    new symbols and emit a Local Symbol Table append with those symbols so that they can be interned use in future
+ *    top-level values.
+ */
+fun interface SymbolInliningStrategy {
+    /**
+     * Represents the different kinds of usage of a symbol token.
+     */
+    enum class SymbolKind {
+        VALUE,
+        FIELD_NAME,
+        ANNOTATION,
+    }
+
+    /**
+     * Indicates whether a particular symbol text should be written inline (as opposed to writing as a SID).
+     */
+    fun shouldWriteInline(symbolKind: SymbolKind, text: String): Boolean
+
```

Comment:
Are we going to call this on every field name, annotation, and symbol value?

Reply:
It does right now. Now that I've been away from this for an hour, it does seem like that could be a performance problem. We can get rid of it if we need to.